Implement Feature types

KeyboardNerd opened the pull request:
The PR is to differentiate packages by their types, and vulnerabilities to affect different types of features.

Change set:
1. Feature Detectors: Extract both binary and source packages from image layers

2. Vulnerability matching algorithm change: Each vulnerability matches a specific type of package, either binary or source package.

| Vulnerability source | Matching Package Type |
|---------------------|---------------|
| Debian: "https://security-tracker.debian.org/tracker" | matching source package name |
| Ubuntu: "https://git.launchpad.net/ubuntu-cve-tracker" | matching source package name |
| Alpine: "https://github.com/alpinelinux/alpine-secdb" | matching binary package name |
| Oracle: "https://linux.oracle.com/oval/" | matching binary package name |
| RHEL: "https://www.redhat.com/security/data/oval/" | matching binary package name |

Please correct the above assumption for each vulnerability data source if any error.

3. API change: Clair returns all found binary and source packages to the client, and let the client to filter.
For each Feature, we add a field to indicate the type of the feature, currently it's Binary or Source package.

KeyboardNerd commented on database/affected_feature_type.go:
```diff
@@ -1,26 +0,0 @@
-// Copyright 2018 clair authors
```
This is renamed to `database/feature_type.go`

KeyboardNerd commented on database/database.go:
```diff
@@ -27,20 +26,20 @@ import (
 	// entity doesn't exist in the database. This error can indicate a wrong
 	// implementation or corrupted database.
-	ErrMissingEntities = errors.New("database: associated immutable entities are missing in the database")
+	ErrMissingEntities = NewStorageError("associated immutable entities are missing in the database")
```
one by one we're moving to use special error type for the database.

KeyboardNerd commented on database/models.go:
```diff
@@ -155,18 +155,33 @@ type Namespace struct {
 	VersionFormat string
 }
 
+func NewNamespace(name string, versionFormat string) *Namespace {
```
we're moving to use the proper initializer for the database structs

KeyboardNerd commented on database/pgsql/complex_test.go:
```diff
@@ -137,65 +132,39 @@ func TestConcurrency(t *testing.T) {
+		t.FailNow()
+	}
 
-	fmt.Printf("%d features, %d vulnerabilities are generated", len(nsFeatures), len(vulnerabilities))
```
Remove the concurrent insertion because the vulnerabilities <-> feature relationships are cached after the features are inserted into the database

KeyboardNerd committed:
pgsql: Add feature_type to initial schema

feature_type is for differentiating the binary packages and source
packages.

KeyboardNerd committed:
database: Add feature_type database model

KeyboardNerd committed:
database: Rename affected Type to feature type

KeyboardNerd committed:
featurefmt: Extract source packages and binary packages
The featurefmt now extracts both binary packages and source packages
from the package manager infos.

KeyboardNerd committed:
database: Update feature model
Remove source name/version fields
Add Type field to indicate if it's binary package or source package

KeyboardNerd committed:
pgsql: Add proper tests for database migration

KeyboardNerd committed:
database: Add StorageError type

KeyboardNerd committed:
pgsql: Fix postgres queries for feature_type

KeyboardNerd committed:
worker: Fix tests for feature_type

KeyboardNerd committed:
Travis: Drop support for postgres 9.4
postgres 9.4 doesn't support ON CONFLICT, which is required in our
implementation.

KeyboardNerd reviewed (commented):
comments for PR reviewer

jzelinskie commented on database/models.go:
```diff
@@ -179,6 +194,11 @@ type NamespacedFeature struct {
 }
 
+func NewNamespacedFeature(namespace *Namespace, feature *Feature) *NamespacedFeature {
+	// TODO: namespaced feature should use pointer values
```
Is it only desirable to use pointers here to save memory?

KeyboardNerd replied:
I think it's idiomatic to use the pointer value for structs

jzelinskie commented on database/pgsql/complex_test.go:
```diff
@@ -137,65 +132,39 @@ func TestConcurrency(t *testing.T) {
-
 func TestCaching(t *testing.T) {
 	store, err := openDatabaseForTest("Caching", false)
 	if !assert.Nil(t, err) {
```
we should just use `require` here instead of assert

jzelinskie commented on database/pgsql/complex_test.go:
```diff
@@ -137,65 +132,39 @@ func TestConcurrency(t *testing.T) {
 	}
 	defer store.Close()
-
 	nsFeatures, vulnerabilities := testGenRandomVulnerabilityAndNamespacedFeature(t, store)
```
I'd actually prefer to keep the newline above this to separate the initialization of the storage from the rest of the setup

KeyboardNerd committed:
api/v3: Add feature type to API feature

API now returns every feature found by the detectors and also indicates
the type of feature.

jzelinskie commented on database/pgsql/complex_test.go:
```diff
@@ -137,65 +132,39 @@ func TestConcurrency(t *testing.T) {
+	tx, err = store.Begin()
 	if !assert.Nil(t, err) {
 		t.FailNow()
 	}
```
I'm pretty sure assert->failnow() is equivalent to just using `require`

jzelinskie commented on database/pgsql/feature.go:
```diff
@@ -240,62 +246,39 @@ func (tx *pgSession) PersistNamespacedFeatures(features []database.NamespacedFea
 	if err != nil {
 		return nil, handleError("searchNamespacedFeaturesVulnerabilities", err)
 	}
+
```
keep the defer in the same block as the query

jzelinskie commented on database/pgsql/migrations/00001_initial_schema.go:
```diff
@@ -159,13 +159,18 @@ var (
 				id SERIAL PRIMARY KEY, 
 				vulnerability_id INT REFERENCES vulnerability ON DELETE CASCADE,
 				feature_name TEXT NOT NULL,
+				feature_type INT NOT NULL REFERENCES feature_type ON DELETE CASCADE,
```
doesn't this mean if the last package of a particular kind gets deleted, so will that entry in the feature_type table?
e.g. if the last source package is deleted `source` ceases to be a package type

KeyboardNerd replied:
I think feature_type here is the parent table, so it's the other way around

KeyboardNerd replied:
https://www.postgresql.org/docs/9.5/ddl-constraints.html
> CASCADE specifies that when a referenced row is deleted, row(s) referencing it should be automatically deleted as well.

row in vulnerability_affected_feature is referencing the feature_type table.
so when a feature type is deleted, we delete the row referencing that.

KeyboardNerd committed:
update according to comments
